File: complex_network_operation_jl/functions/fun_generate_points.jl
"""
函数：生成散点分布
"""

using Random
using LinearAlgebra
using Printf

include("../../utils/altorithm_utils.jl")

function generate_points(
        set_densityDistance::Union{Float64, Nothing}=nothing,
        set_numPoints::Union{Int, Nothing}=nothing,
        set_circleRadius::Union{Float64, Nothing}=nothing,
        circle_origin::Vector{Float64}=[0.0, 0.0],
        num_samples::Int=30,
        distribution::String="Normal",
        algorithm::Union{String, Nothing}=nothing,
        sigma::Float64=1.0,
        hierarchical_network_type::Bool=false,
        num_external_nodes::Int=0
    )
    """
    根据一系列初始条件与给定分布生成符合要求的散点。

    这里估算的过程中，采用了正方形、立方体作为中间计算量，然后转换为同面积、同体积的圆形、球体。

    散点分布类型：
    - 'Posssion Disk': 生成符合 Poisson Disk 分布的散点。这种分布的特点是，散点之间的距离大致相等，且散点之间的距离大于给定的最小距离。
    - 'Uniform': 生成符合均匀分布的散点。这种分布的特点是，散点之间的距离大致相等，但散点之间的距离不一定大于给定的最小距离。
    - 'Normal': 生成符合正态分布的散点。这种分布的特点是，散点之间的距离大致符合正态分布，但散点之间的距离不一定大于给定的最小距离。

    散点分布相关的算法类型：
    - 'Bridson': 采用 Bridson 算法生成符合 Poisson Disk 分布的散点。该算法的特点是生成的散点距离一定满足不小于最小距离，但是生成的散点数量不一定等于给定的数量；
    - 'K-means': 采用 K-means 算法生成符合接近 Poisson Disk 分布的散点。该算法的特点是能够生成精确数量的散点，并且生成性能较高；

    Args:
        set_densityDistance: Float64, 密度距离
        set_numPoints: Int, 散点之总数。注意实际生成的数量大约在这个值附近，不会完全等于这个值。
        set_circleRadius: Float64, 范围半径
        circle_origin: Vector{Float64}, 圆心坐标
        num_samples: Int, 算法每次迭代生成的散点数量
        distribution: String, 期望的散点分布类型
        algorithm: Union{String, Nothing}, 生成散点分布之算法
        sigma: Float64, 正态分布的标准差
        hierarchical_network_type: Bool, 是否是层次网络
        num_external_nodes: Int, 外部节点的数量

    Returns:
        Matrix{Float64}: 生成的散点
    """
    # 根据给定参数计算其余参数
    # 判断生成的空间维度
    dim = length(circle_origin)
    if dim == 2
        if set_densityDistance !== nothing && set_numPoints !== nothing && set_circleRadius === nothing
            a = set_densityDistance * sqrt(set_numPoints)
            circle_radius = a / sqrt(π)
        elseif set_densityDistance !== nothing && set_circleRadius !== nothing && set_numPoints === nothing
            a = set_circleRadius * sqrt(π)
            num_points = round(Int, a / set_densityDistance)
        elseif set_numPoints !== nothing && set_circleRadius !== nothing && set_densityDistance === nothing
            a = set_circleRadius * sqrt(π)
            density_distance = a / set_numPoints
        else
            throw(ArgumentError("Two out of the three parameters (density, total, area) must be provided."))
        end
    else
        throw(ArgumentError("The dimension of the circle origin must be either 2."))
    end

    println("dim: $dim, density_distance: $density_distance, num_points: $num_points, circle_radius: $circle_radius, circle_origin: $circle_origin")

    # 生成散点
    if hierarchical_network_type
        return _generate_hierarchical_network_nodes(circle_origin, circle_radius, num_points, num_external_nodes)
    end

    if distribution == "Normal"
        return _generate_points_used_Normal_random_distribution(circle_origin, circle_radius, num_points, sigma)
    elseif distribution == "Uniform"
        return _generate_points_used_Uniform_random_distribution(circle_origin, circle_radius, num_points)
    elseif distribution == "Posssion Disk"
        if algorithm == "K-means"
            return _generate_points_used_PoissonDisk_random_distribution_by_Kmeans_algorithm(circle_origin, circle_radius, num_points)
        elseif algorithm == "Bridson"
            return _generate_points_used_PoissonDisk_random_distribution_by_Bridson_algorithm(circle_origin, circle_radius, density_distance, num_samples)
        elseif algorithm === nothing
            if set_numPoints !== nothing
                algorithm = "K-means"
                return _generate_points_used_PoissonDisk_random_distribution_by_Kmeans_algorithm(circle_origin, circle_radius, num_points)
            elseif set_densityDistance !== nothing && set_numPoints === nothing && set_circleRadius !== nothing
                algorithm = "Bridson"
                return _generate_points_used_PoissonDisk_random_distribution_by_Bridson_algorithm(circle_origin, circle_radius, density_distance, num_samples)
            end
        end
    end

    return nothing
end

function _generate_points_used_Normal_random_distribution(circle_origin::Vector{Float64}, circle_radius::Float64, num_points::Int, sigma::Float64=1.0)
    """
    生成符合正态分布的散点。这里采用了简单的方法，即在圆内随机生成一定数量的点。

    Args:
        circle_origin: Vector{Float64}, 圆心坐标
        circle_radius: Float64, 圆的半径
        num_points: Int, 生成的点的数量
        sigma: Float64, 正态分布的标准差

    Returns:
        Matrix{Float64}: 生成的散点
    """
    points = zeros(Float64, num_points, 2)
    for i in 1:num_points
        theta = rand() * 2 * π
        r = sqrt(rand()) * circle_radius
        points[i, :] = [r * cos(theta) + circle_origin[1], r * sin(theta) + circle_origin[2]]
    end
    return points
end

function _generate_points_used_Uniform_random_distribution(circle_origin::Vector{Float64}, circle_radius::Float64, num_points::Int)
    """
    生成符合均匀分布的散点。这里采用了简单的方法，即在圆内随机生成一定数量的点。

    Args:
        circle_origin: Vector{Float64}, 圆心坐标
        circle_radius: Float64, 圆的半径
        num_points: Int, 生成的点的数量

    Returns:
        Matrix{Float64}: 生成的散点
    """
    points = zeros(Float64, num_points, 2)
    for i in 1:num_points
        theta = rand() * 2 * π
        r = sqrt(rand()) * circle_radius
        points[i, :] = [r * cos(theta) + circle_origin[1], r * sin(theta) + circle_origin[2]]
    end
    return points
end

function _generate_points_used_PoissonDisk_random_distribution_by_Bridson_algorithm(origin::Vector{Float64}, radius::Float64, min_distance::Float64, num_samples::Int=15)
    """
    生成符合 Poisson Disk 分布的散点，这些散点分布在一个圆形区域内。这里采用了著名的 Bridson 算法。

    Args:
        origin: Vector{Float64}, 圆心坐标
        radius: Float64, 圆的半径
        min_distance: Float64, 最小距离
        num_samples: Int, 每次生成的散点数量

    Returns:
        Matrix{Float64}: 生成的散点
    """
    cell_size = min_distance / sqrt(2)
    grid_radius = Int(radius / cell_size) + 1
    grid = fill(-1, 2 * grid_radius, 2 * grid_radius)
    active_list = []

    samples = zeros(Float64, 4 * radius * radius, 2)
    sample_count = 0

    first_sample = [rand(-radius:radius), rand(-radius:radius)]
    while norm(first_sample) > radius
        first_sample = [rand(-radius:radius), rand(-radius:radius)]
    end

    samples[sample_count + 1, :] = first_sample
    sample_count += 1
    grid_index = (Int((first_sample[1] + radius) / cell_size), Int((first_sample[2] + radius) / cell_size))
    grid[grid_index...] = 0
    push!(active_list, grid_index)

    while !isempty(active_list)
        current_index = rand(1:length(active_list))
        current_cell = active_list[current_index]

        for _ in 1:num_samples
            new_sample = _generate_random_point_around_by_Bridson_algorithm(samples[grid[current_cell...], :], min_distance)
            if norm(new_sample) <= radius && _is_valid_sample_by_Bridson_algorithm(new_sample, radius, min_distance, samples[1:sample_count, :], grid, cell_size)
                samples[sample_count + 1, :] = new_sample
                sample_count += 1
                new_index = (Int((new_sample[1] + radius) / cell_size), Int((new_sample[2] + radius) / cell_size))
                grid[new_index...] = sample_count - 1
                push!(active_list, new_index)
            end
        end

        deleteat!(active_list, current_index)
    end

    samples = samples[1:sample_count, :] .+ origin
    return samples
end

function _generate_random_point_around_by_Bridson_algorithm(point::Vector{Float64}, min_distance::Float64)
    """
    生成一个距离 point 一定距离的随机点。

    Args:
        point: Vector{Float64}, 中心点
        min_distance: Float64, 最小距离

    Returns:
        Vector{Float64}: 生成的随机点
    """
    r = rand(min_distance:2 * min_distance)
    theta = rand() * 2 * π
    x = point[1] + r * cos(theta)
    y = point[2] + r * sin(theta)
    return [x, y]
end

function _is_valid_sample_by_Bridson_algorithm(sample::Vector{Float64}, radius::Float64, min_distance::Float64, samples::Matrix{Float64}, grid::Matrix{Int}, cell_size::Float64)
    """
    判断生成的散点是否有效。散点必须满足以下条件：
    1. 位于空间内
    2. 与其他散点的距离大于 min_distance
    3. 与邻居散点的距离大于 min_distance

    Args:
        sample: Vector{Float64}, 生成的散点
        radius: Float64, 圆的半径
        min_distance: Float64, 最小距离
        samples: Matrix{Float64}, 已生成的散点
        grid: Matrix{Int}, 网格
        cell_size: Float64, 网格的大小

    Returns:
        Bool: 是否有效
    """
    if norm(sample) > radius
        return false
    end
    grid_index = (Int((sample[1] + radius) / cell_size), Int((sample[2] + radius) / cell_size))
    if grid_index[1] < 1 || grid_index[1] > size(grid, 1) || grid_index[2] < 1 || grid_index[2] > size(grid, 2)
        return false
    end
    if grid[grid_index...] != -1
        return false
    end
    for neighbor_index in _get_neighbor_indices(grid_index, size(grid))
        if neighbor_index[1] >= 1 && neighbor_index[1] <= size(grid, 1) && neighbor_index[2] >= 1 && neighbor_index[2] <= size(grid, 2) && grid[neighbor_index...] != -1
            neighbor_sample = samples[grid[neighbor_index...], :]
            if norm(sample .- neighbor_sample) < min_distance
                return false
            end
        end
    end
    return true
end

function _get_neighbor_indices(index::Tuple{Int, Int}, grid_shape::Tuple{Int, Int})
    """
    获取邻居散点的索引
    """
    neighbors = []
    for i in -1:1
        for j in -1:1
            if i == 0 && j == 0
                continue
            end
            neighbor_index = (index[1] + i, index[2] + j)
            if neighbor_index[1] >= 1 && neighbor_index[1] <= grid_shape[1] && neighbor_index[2] >= 1 && neighbor_index[2] <= grid_shape[2]
                push!(neighbors, neighbor_index)
            end
        end
    end
    return neighbors
end

function _generate_points_used_PoissonDisk_random_distribution_by_Kmeans_algorithm(circle_origin::Vector{Float64}, circle_radius::Float64, num_points::Int)
    """
    生成符合接近 Poisson Disk 分布的散点。这里采用了 K-means 算法。

    Args:
        num_points: Int, 目标生成的总点数
        circle_origin: Vector{Float64}, 圆心坐标
        circle_radius: Float64, 圆的半径

    Returns:
        Matrix{Float64}: 生成的散点
    """
    num_initial_points = num_points * 20
    initial_points = zeros(Float64, num_initial_points, 2)
    for i in 1:num_initial_points
        theta = rand() * 2 * π
        r = sqrt(rand()) * circle_radius
        initial_points[i, :] = [r * cos(theta) + circle_origin[1], r * sin(theta) + circle_origin[2]]
    end

    centroids = kmeans(initial_points, num_points)

    return centroids
end

function _generate_hierarchical_network_nodes(circle_origin::Vector{Float64}, circle_radius::Float64, num_points::Int, num_external_nodes::Int)
    """
    生成层次网络节点，包括内部节点和外部节点。

    Args:
        circle_origin: Vector{Float64}, 圆心坐标
        circle_radius: Float64, 圆的半径
        num_points: Int, 生成的点的数量
        num_external_nodes: Int, 外部节点的数量

    Returns:
        Matrix{Float64}: 生成的散点
    """
    points = zeros(Float64, num_points, 2)
    external_points = zeros(Float64, num_external_nodes, 2)
    internal_points = zeros(Float64, num_points - num_external_nodes, 2)

    for i in 1:num_external_nodes
        theta = rand() * 2 * π
        r = circle_radius
        external_points[i, :] = [r * cos(theta) + circle_origin[1], r * sin(theta) + circle_origin[2]]
    end

    for i in 1:(num_points - num_external_nodes)
        theta = rand() * 2 * π
        r = sqrt(rand()) * circle_radius
        internal_points[i, :] = [r * cos(theta) + circle_origin[1], r * sin(theta) + circle_origin[2]]
    end

    points[1:num_external_nodes, :] = external_points
    points[(num_external_nodes + 1):end, :] = internal_points

    return points
end

File: complex_network_operation_js/functions/fun_generate_points.js
import * as d3 from 'd3';

/**
 * 函数：生成散点分布
 */



/**
 * 根据一系列初始条件与给定分布生成符合要求的散点。
 * 
 * 这里估算的过程中，采用了正方形、立方体作为中间计算量，然后转换为同面积、同体积的圆形、球体。
 * 
 * 散点分布类型：
 * - 'Posssion Disk': 生成符合 Poisson Disk 分布的散点。这种分布的特点是，散点之间的距离大致相等，且散点之间的距离大于给定的最小距离。
 * - 'Uniform': 生成符合均匀分布的散点。这种分布的特点是，散点之间的距离大致相等，但散点之间的距离不一定大于给定的最小距离。
 * - 'Normal': 生成符合正态分布的散点。这种分布的特点是，散点之间的距离大致符合正态分布，但散点之间的距离不一定大于给定的最小距离。
 * 
 * 散点分布相关的算法类型：
 * - 'Bridson': 采用 Bridson 算法生成符合 Poisson Disk 分布的散点。该算法的特点是生成的散点距离一定满足不小于最小距离，但是生成的散点数量不一定等于给定的数量；
 * - 'cKDTree': 采用 cKDTree 算法生成符合 Poisson Disk 分布的散点。该算法的特点是生成的散点距离一定满足不小于最小距离，但是生成的散点数量不一定等于给定的数量；#BUG 这个还没有适配生成圆形边界的情况。因此不要使用！
 * - 'K-means': 采用 K-means 算法生成符合接近 Poisson Disk 分布的散点。该算法的特点是能够生成精确数量的散点，并且生成性能较高；
 * - 'annealing': 采用退火算法生成符合 Poisson Disk 分布的散点。该算法的特点是能够生成精确数量的散点，但是生成性能较差； #BUG 还没有测试。因此不要使用！
 * 
 * @param {number} set_densityDistance - 密度距离
 * @param {number} set_numPoints - 散点总数。注意实际生成的数量大约在这个值附近，不会完全等于这个值。
 * @param {number} set_circleRadius - 范围半径
 * @param {Array} circle_origin - 圆心坐标
 * @param {number} num_samples - 每次迭代生成的散点数量
 * @param {string} distribution - 期望的散点分布类型
 * @param {string} algorithm - 生成散点分布的算法
 * @param {number} sigma - 正态分布的标准差
 * @param {boolean} hierarchical_network_type - 是否是层次网络
 * @param {number} num_external_nodes - 外部节点的数量
 * 
 * @example
 * const points = generatePoints({ set_numPoints: 100, set_circleRadius: 500, circle_origin: [0, 0], distribution: 'Normal' });
 * console.log(points);
 * 
 * @returns {Array} 生成的散点。散点数据结构形式为：[[x1, y1], [x2, y2], ...]
 */
function generatePoints(setDensityDistance = null, setNumPoints = null, setCircleRadius = null, circleOrigin = [0, 0], numSamples = 30, distribution = 'Normal', algorithm = null, sigma = 1, hierarchicalNetworkType = false, numExternalNodes = 0) {
    let dim = circleOrigin.length;
    let densityDistance, numPoints, circleRadius;

    if (dim !== 2) {
        throw new Error("The dimension of the circle origin must be 2.");
    }

    if (setDensityDistance !== null && setNumPoints !== null && setCircleRadius === null) {
        let a = setDensityDistance * Math.sqrt(setNumPoints);
        circleRadius = a / Math.sqrt(Math.PI);
        densityDistance = setDensityDistance;
        numPoints = setNumPoints;
    } else if (setDensityDistance !== null && setCircleRadius !== null && setNumPoints === null) {
        let a = setCircleRadius * Math.sqrt(Math.PI);
        numPoints = Math.round(a / setDensityDistance);
        densityDistance = setDensityDistance;
        circleRadius = setCircleRadius;
    } else if (setNumPoints !== null && setCircleRadius !== null && setDensityDistance === null) {
        let a = setCircleRadius * Math.sqrt(Math.PI);
        densityDistance = a / setNumPoints;
        numPoints = setNumPoints;
        circleRadius = setCircleRadius;
    } else {
        throw new Error("Two out of the three parameters (density, total, area) must be provided.");
    }

    if (hierarchicalNetworkType) {
        return generateHierarchicalNetworkNodes(circleOrigin, circleRadius, numPoints, numExternalNodes);
    }

    switch (distribution) {
        case 'Normal':
            return generatePointsUsingNormalRandomDistribution(circleOrigin, circleRadius, numPoints, sigma);
        case 'Uniform':
            return generatePointsUsingUniformRandomDistribution(circleOrigin, circleRadius, numPoints);
        case 'Poisson Disk':
            switch (algorithm) {
                case 'K-means':
                    return generatePointsUsingPoissonDiskRandomDistributionByKmeansAlgorithm(circleOrigin, circleRadius, numPoints);
                case 'Bridson':
                    return generatePointsUsingPoissonDiskRandomDistributionByBridsonAlgorithm(circleOrigin, circleRadius, densityDistance, numSamples);
                default:
                    if (setNumPoints !== null) {
                        return generatePointsUsingPoissonDiskRandomDistributionByKmeansAlgorithm(circleOrigin, circleRadius, numPoints);
                    } else if (setDensityDistance !== null && setNumPoints === null && setCircleRadius !== null) {
                        return generatePointsUsingPoissonDiskRandomDistributionByBridsonAlgorithm(circleOrigin, circleRadius, densityDistance, numSamples);
                    }
            }
    }
}

/**
 * 生成符合正态分布的散点。
 * 
 * @param {Array} circle_origin - 圆心坐标
 * @param {number} circle_radius - 圆的半径
 * @param {number} num_points - 生成的点的数量
 * @param {number} sigma - 正态分布的标准差
 * @returns {Array} 生成的散点
 */
function generatePointsUsingNormalRandomDistribution(circleOrigin, circleRadius, numPoints, sigma = 1) {
    const points = [];
    for (let i = 0; i < numPoints; i++) {
        const theta = Math.random() * 2 * Math.PI;
        const r = Math.sqrt(Math.random()) * circleRadius;
        points.push([r * Math.cos(theta) + circleOrigin[0], r * Math.sin(theta) + circleOrigin[1]]);
    }
    return points;
}

/**
 * 生成符合均匀分布的散点。
 * 
 * @param {Array} circle_origin - 圆心坐标
 * @param {number} circle_radius - 圆的半径
 * @param {number} num_points - 生成的点的数量
 * @returns {Array} 生成的散点
 */
function generatePointsUsingUniformRandomDistribution(circleOrigin, circleRadius, numPoints) {
    const points = [];
    for (let i = 0; i < numPoints; i++) {
        const theta = Math.random() * 2 * Math.PI;
        const r = Math.sqrt(Math.random()) * circleRadius;
        points.push([r * Math.cos(theta) + circleOrigin[0], r * Math.sin(theta) + circleOrigin[1]]);
    }
    return points;
}

/**
 * 采用 Bridson 算法生成符合 Poisson Disk 分布的散点。
 * 
 * @param {Array} origin - 圆心坐标
 * @param {number} radius - 圆的半径
 * @param {number} min_distance - 最小距离
 * @param {number} num_samples - 每次迭代生成的散点数量
 * @returns {Array} 生成的散点
 */
function generatePointsUsingPoissonDiskRandomDistributionByBridsonAlgorithm(origin, radius, minDistance, numSamples = 15) {
    const cellSize = minDistance / Math.sqrt(2);
    const gridRadius = Math.floor(radius / cellSize) + 1;
    const grid = Array.from({ length: 2 * gridRadius }, () => Array(2 * gridRadius).fill(-1));
    const activeList = [];
    const samples = [];
    let sampleCount = 0;

    let firstSample = [Math.random() * 2 * radius - radius, Math.random() * 2 * radius - radius];
    while (Math.hypot(firstSample[0], firstSample[1]) > radius) {
        firstSample = [Math.random() * 2 * radius - radius, Math.random() * 2 * radius - radius];
    }

    samples.push(firstSample);
    sampleCount++;
    const gridIndex = [Math.floor((firstSample[0] + radius) / cellSize), Math.floor((firstSample[1] + radius) / cellSize)];
    grid[gridIndex[0]][gridIndex[1]] = 0;
    activeList.push(gridIndex);

    while (activeList.length > 0) {
        const currentIndex = Math.floor(Math.random() * activeList.length);
        const currentCell = activeList[currentIndex];

        for (let i = 0; i < numSamples; i++) {
            const newSample = generateRandomPointAroundByBridsonAlgorithm(samples[grid[currentCell[0]][currentCell[1]]], minDistance);
            if (Math.hypot(newSample[0], newSample[1]) <= radius && isValidSampleByBridsonAlgorithm(newSample, radius, minDistance, samples, grid, cellSize)) {
                samples.push(newSample);
                sampleCount++;
                const newIndex = [Math.floor((newSample[0] + radius) / cellSize), Math.floor((newSample[1] + radius) / cellSize)];
                grid[newIndex[0]][newIndex[1]] = sampleCount - 1;
                activeList.push(newIndex);
            }
        }

        activeList.splice(currentIndex, 1);
    }

    return samples.map(sample => [sample[0] + origin[0], sample[1] + origin[1]]);
}

/**
 * 采用 Bridson 算法生成一个距离给定点一定距离的随机点。
 * 
 * @param {Array} point - 中心点
 * @param {number} min_distance - 最小距离
 * @returns {Array} 生成的随机点
 */
function generateRandomPointAroundByBridsonAlgorithm(point, minDistance) {
    const r = Math.random() * minDistance + minDistance;
    const theta = Math.random() * 2 * Math.PI;
    return [point[0] + r * Math.cos(theta), point[1] + r * Math.sin(theta)];
}

/**
 * 判断生成的散点是否有效。
 * 
 * @param {Array} sample - 生成的散点
 * @param {number} radius - 圆的半径
 * @param {number} min_distance - 最小距离
 * @param {Array} samples - 已生成的散点
 * @param {Array} grid - 网格
 * @param {number} cell_size - 网格的大小
 * @returns {boolean} 是否有效
 */
function isValidSampleByBridsonAlgorithm(sample, radius, minDistance, samples, grid, cellSize) {
    if (Math.hypot(sample[0], sample[1]) > radius) {
        return false;
    }
    const gridIndex = [Math.floor((sample[0] + radius) / cellSize), Math.floor((sample[1] + radius) / cellSize)];
    if (gridIndex[0] < 0 || gridIndex[0] >= grid.length || gridIndex[1] < 0 || gridIndex[1] >= grid[0].length) {
        return false;
    }
    if (grid[gridIndex[0]][gridIndex[1]] !== -1) {
        return false;
    }
    for (const neighborIndex of getNeighborIndices(gridIndex, grid.length, grid[0].length)) {
        if (grid[neighborIndex[0]][neighborIndex[1]] !== -1) {
            const neighborSample = samples[grid[neighborIndex[0]][neighborIndex[1]]];
            if (Math.hypot(sample[0] - neighborSample[0], sample[1] - neighborSample[1]) < minDistance) {
                return false;
            }
        }
    }
    return true;
}

/**
 * 获取邻居散点的索引。
 * 
 * @param {Array} index - 索引
 * @param {number} grid_width - 网格宽度
 * @param {number} grid_height - 网格高度
 * @returns {Array} 邻居索引
 */
function getNeighborIndices(index, gridWidth, gridHeight) {
    const neighbors = [];
    for (let i = -1; i <= 1; i++) {
        for (let j = -1; j <= 1; j++) {
            if (i === 0 && j === 0) continue;
            const neighborIndex = [index[0] + i, index[1] + j];
            if (neighborIndex[0] >= 0 && neighborIndex[0] < gridWidth && neighborIndex[1] >= 0 && neighborIndex[1] < gridHeight) {
                neighbors.push(neighborIndex);
            }
        }
    }
    return neighbors;
}

/**
 * 采用 K-means 算法生成符合 Poisson Disk 分布的散点。
 * 
 * @param {Array} circle_origin - 圆心坐标
 * @param {number} circle_radius - 圆的半径
 * @param {number} num_points - 生成的点的数量
 * @returns {Array} 生成的散点
 */
function generatePointsUsingPoissonDiskRandomDistributionByKmeansAlgorithm(circleOrigin, circleRadius, numPoints) {
    const numInitialPoints = numPoints * 20;
    const initialPoints = [];
    for (let i = 0; i < numInitialPoints; i++) {
        const theta = Math.random() * 2 * Math.PI;
        const r = Math.sqrt(Math.random()) * circleRadius;
        initialPoints.push([r * Math.cos(theta) + circleOrigin[0], r * Math.sin(theta) + circleOrigin[1]]);
    }

    const centroids = d3.cluster().kmeans(initialPoints, numPoints);
    return centroids;
}

/**
 * 实现 K-means 算法。
 * 
 * @param {Array} points - 二维数组，每一行代表一个点。
 * @param {number} num_clusters - 聚类的数量。
 * @param {number} max_iter - 最大迭代次数。
 * @returns {Array} 每一行代表一个质心。
 */
function kmeans(points, num_clusters, max_iter = 100) {
    const num_points = points.length;
    const num_dim = points[0].length;
    let centroids = points.slice(0, num_clusters);

    for (let iter = 0; iter < max_iter; iter++) {
        const distances = Array.from({ length: num_points }, () => Array(num_clusters).fill(0));
        for (let i = 0; i < num_points; i++) {
            for (let j = 0; j < num_clusters; j++) {
                distances[i][j] = points[i].reduce((sum, val, dim) => sum + Math.pow(val - centroids[j][dim], 2), 0);
            }
        }

        const labels = distances.map(row => row.indexOf(Math.min(...row)));

        const new_centroids = Array.from({ length: num_clusters }, () => Array(num_dim).fill(0));
        const counts = Array(num_clusters).fill(0);
        for (let i = 0; i < num_points; i++) {
            const label = labels[i];
            counts[label]++;
            for (let dim = 0; dim < num_dim; dim++) {
                new_centroids[label][dim] += points[i][dim];
            }
        }

        for (let j = 0; j < num_clusters; j++) {
            for (let dim = 0; dim < num_dim; dim++) {
                new_centroids[j][dim] /= counts[j] || 1;
            }
        }

        if (centroids.every((centroid, j) => centroid.every((val, dim) => val === new_centroids[j][dim]))) {
            break;
        }
        centroids = new_centroids;
    }

    return centroids;
}

/**
 * 生成层次网络节点。
 * 
 * @param {Array} circle_origin - 圆心坐标
 * @param {number} circle_radius - 圆的半径
 * @param {number} num_points - 生成的点的数量
 * @param {number} num_external_nodes - 外部节点的数量
 * @returns {Array} 生成的散点
 */
function generateHierarchicalNetworkNodes(circleOrigin, circleRadius, numPoints, numExternalNodes) {
    const points = [];
    const externalPoints = [];
    const internalPoints = [];

    for (let i = 0; i < numExternalNodes; i++) {
        const theta = Math.random() * 2 * Math.PI;
        const r = circleRadius;
        externalPoints.push([r * Math.cos(theta) + circleOrigin[0], r * Math.sin(theta) + circleOrigin[1]]);
    }

    for (let i = 0; i < numPoints - numExternalNodes; i++) {
        const theta = Math.random() * 2 * Math.PI;
        const r = Math.sqrt(Math.random()) * circleRadius;
        internalPoints.push([r * Math.cos(theta) + circleOrigin[0], r * Math.sin(theta) + circleOrigin[1]]);
    }

    return externalPoints.concat(internalPoints);
}

export { generatePoints };